src/prommis/nanofiltration/costing/diafiltration.py

```diff
@@ -0,0 +1,713 @@
+    m.fs.stage3.retentate_side_stream_state[0, 10].conc_mass_solute["Co"].fix(C_Co_feed)
+
+    assert degrees_of_freedom(m) == 0
+
+    # Initialize Mixer 2
+    propagate_state(
+        destination=m.fs.mix2.inlet_2,
+        source=m.fs.stage2.permeate_outlet,
+    )
+
+    mix_initializer.initialize(m.fs.mix2)
+
+    # Try to initialize stage 3
+    propagate_state(
+        destination=m.fs.stage3.retentate_inlet,
+        source=m.fs.mix2.outlet,
+    )
+
+    initializer.initialize(m.fs.stage3)
+
+    # Lets try again
+    # Initial feed guess is pure diafiltrate (no recycle)
+    # Remember to only set the value, not fix the inlet
+    m.fs.stage3.retentate_inlet.flow_vol[0].set_value(Q_diaf)
+    m.fs.stage3.retentate_inlet.conc_mass_solute[0, "Li"].set_value(C_Li_diaf)
+    m.fs.stage3.retentate_inlet.conc_mass_solute[0, "Co"].set_value(C_Co_diaf)
+
+    initializer.initialize(m.fs.stage3)
+
+    add_expressions(m)
+
+    return m
+
+
+def create_rules(m):
+    def solvent_rule(b, s):
+        return (
+            b.material_transfer_term[0, s, "permeate", "retentate", "solvent"]
+            == Jw * b.length * w * m.fs.properties.dens_H2O / 10
+        )
+
+    def solute_rule(b, s, j):
+        if s == 1:
+            in_state = b.retentate_inlet_state[0]
+        else:
+            sp = b.elements.prev(s)
+            in_state = b.retentate[0, sp]
+
+        return pyo.log(b.retentate[0, s].conc_mass_solute[j]) + (
+            m.fs.sieving_coefficient[j] - 1
+        ) * pyo.log(in_state.flow_vol) == pyo.log(in_state.conc_mass_solute[j]) + (
+            m.fs.sieving_coefficient[j] - 1
+        ) * pyo.log(
+            b.retentate[0, s].flow_vol
+        )
+
+    def stage3_solute_rule(b, s, j):
+        if s == 1:
+            q_in = b.retentate_inlet_state[0].flow_vol
+            c_in = b.retentate_inlet_state[0].conc_mass_solute[j]
+        elif s == 10:
+            sp = b.elements.prev(s)
+            q_in = (
+                b.retentate[0, sp].flow_vol
+                + b.retentate_side_stream_state[0, 10].flow_vol
+            )
+            c_in = (
+                b.retentate[0, sp].conc_mass_solute[j] * b.retentate[0, sp].flow_vol
+                + b.retentate_side_stream_state[0, 10].conc_mass_solute[j]
+                * b.retentate_side_stream_state[0, 10].flow_vol
+            ) / q_in
+        else:
+            sp = b.elements.prev(s)
+            q_in = b.retentate[0, sp].flow_vol
+            c_in = b.retentate[0, sp].conc_mass_solute[j]
+
+        return pyo.log(b.retentate[0, s].conc_mass_solute[j]) + (
+            m.fs.sieving_coefficient[j] - 1
+        ) * pyo.log(q_in) == pyo.log(c_in) + (
+            m.fs.sieving_coefficient[j] - 1
+        ) * pyo.log(
+            b.retentate[0, s].flow_vol
+        )
+
+    return (solvent_rule, solute_rule, stage3_solute_rule)
+
+
+def solve_model(m):
+    solver = pyo.SolverFactory("ipopt")
```

Comment:
I recommend trying to use `get_solver` from IDAES if possible. For one, this means we have a standard solver across all code, and the default solver is now the new `"ipopt_v2"` interface with a linear presolve (and in future will also include model scaling by default).

Using `get_solver` means that we do not need to go change every instance of solver creation when we make changes to our solvers.